common/utilities/string/wrap.cpp

```diff
@@ -0,0 +1,95 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2020 Intel Corporation. All Rights Reserved.
+
+#include <string>
+#include <vector>
+#include <sstream>
+#include "wrap.h"
+#include "split.h"
+#include "../third-party/imgui/imgui.h"
+
+namespace utilities {
+namespace string {
+
+std::string wrap_paragraph( const std::string & paragraph, int wrap_pixels_width )
+{
+    float space_size = ImGui::CalcTextSize( " " ).x;  // Calculate space width in pixels
+    std::string wrapped_line;                         // The line that is wrapped in this iteration
+    std::string wrapped_paragraph;                    // The output wrapped paragraph
+    auto remaining_paragraph = paragraph;             // holds the remaining unwrapped part of the input paragraph
+    auto next_word = remaining_paragraph.substr(
+        0,
+        remaining_paragraph.find( " " ) );  // The next word to add to the current line
+    bool first_word = true;
+
+    while( ! next_word.empty() )
+    {
+        float next_x = 0.0f;
+        // If this is the first word we try to place it first in line,
+        // if not we concatenate it to the last word after adding a space
+        if( ! first_word )
+        {
+            next_x = ImGui::CalcTextSize( wrapped_line.c_str() ).x + space_size;
+        }
+
+        if( next_x + ImGui::CalcTextSize( next_word.c_str() ).x <= wrap_pixels_width )
+        {
+            if( ! first_word )
+                wrapped_line += " ";  // first word should not start with " "
+            wrapped_line += next_word;
+        }
+        else
+        {  // current line cannot feat new word so we wrap the line and
+           // start building the new line
+
+            wrapped_paragraph += wrapped_line;  // copy line build by now
+            if( ! first_word )
+                wrapped_paragraph += '\n';      // break the previous line if exist
+            wrapped_line = next_word;           // add next work to new line
+        }
+
+        first_word = false;
+
+        // If we have more characters left other then the current word, prepare inputs for next
+        // iteration, If not add the current line built so far to the output and finish current
+        // line wrap.
+        if( remaining_paragraph.size() > next_word.size() )
+        {
+            remaining_paragraph = remaining_paragraph.substr( next_word.size() + 1 );
```

Comment:
What if there's more than one space?

Reply:
Need to add checks
CHECK( wrap( "ab  cd", 32 ) == "ab\ncd" );
CHECK( wrap( "ab cd   ", 32 ) == "ab\ncd" );